lib/queue.js

```diff
@@ -673,20 +673,20 @@ class Queue extends Emitter {
             return;
           }
 
-          this.running += 1;
-          this.queued -= 1;
-          if (this.running + this.queued < this.concurrency) {
-            this.queued += 1;
-            setImmediate(jobTick);
-          }
-
           if (!job) {
             // Per comment in Queue#_waitForJob, this branch is possible when
             // the job is removed before processing can take place, but after
             // being initially acquired.
             return;
           }
 
+          this.running += 1;
+          this.queued -= 1;
+          if (this.running + this.queued < this.concurrency) {
+            this.queued += 1;
+            setImmediate(jobTick);
+          }
+
```

Comment:
What's the intent behind moving these lines? Seems like we're interfering with the intended function of perpetuating one logical "thread"/worker to read and process jobs by doing this.

Reply:
The previous code was (or at least seems) logically wrong -- when `job` is falsey, we immediately stop this worker and never decrement `this.running` (which happens on line 691 below), so it grows without bound.

As it happens we only ever check the sum `this.running + this.queued` so the combination of decrementing one and incrementing the other actually has no practical effect anyway, which I guess is the reason why this "bug" is in the released code.

Reply:
I guess there is also a control flow change inherent in moving this code block. (the `setImmediate(jobTick)`). 

We are about to exit this promise anyway, which will call `setImmediate(jobTick)` in the `finally` block. Assuming the next job we call isn't missing too, we'll start queuing up new workers if there's concurrency available so it isn't a big deal, but if you are concerned we could just do the reincrement/redecrement in the `if (!job)` instead.

Reply:
urgh I hate how this code is structured. every time I read it I think I spot something wrong only to figure out I'm mistaken. if we add a test that fails before this PR and succeeds after it regarding spoolup then I'm cool with it :)

Reply:
Any hints on how to write this test? Should I create a bunch of jobs, remove the data for one (manually via a redis call) and then start the queue w/ some concurrency?

Reply:
yeah that seems like a good start. I'd also block all jobs in some appropriate manner and wait for the appropriate number to start to make sure the correct number of worker "threads" run.